drivers: ethernet: eth_sam_gmac: Fix priority queues

nandojve opened the pull request:
The Atmel SAM SoC with ethernet port uses same GMAC driver. However, there are differences between SoC GMAC implementation. Some SoCs have priority queue and system can configure 0 up to 5, depending of SoC version. This update current GMAC driver adding missing definitions.

I already tested with SAM_V71_XULT using 6 queues [5 pri queues] with samples/net/telnet with DHCP. Ping is working OK using telnet terminal. It needs people to test SAME70[B]. I already tested SAM_V71_XULT as SAMV71**A** variation and seems to works well.

I added Kconfig rules to make sure user can only select proper queue size.

NOTE for SAM4E: Not yet supported.
NOTE for SAME70[B]: https://github.com/zephyrproject-rtos/hal_atmel/pull/7

This work is based on slack discussion this week and the start point was https://github.com/zephyrproject-rtos/zephyr/pull/12711.

This PR supersedes #12711 (closes #12711).

dgloeck committed:
drivers: eth: sam-e70: revision B has more queues

The first revision of the SAM E70 soc had three queues. The current
revision B has six queues. If we don't initialize all queues, the DMA
engine gets stuck when trying to read a descriptor from NULL. To enable
the initialization of the additional queues, the correct soc has to be
selected in the config options, f.ex. CONFIG_SOC_PART_NUMBER_SAME70Q21B
instead of CONFIG_SOC_PART_NUMBER_SAME70Q21.

Also rename GMAC_QUEUE_NO to GMAC_QUEUE_NUM as requested during review.

Signed-off-by: Daniel Glöckner <dg@emlix.com>

nandojve committed:
drivers: eth: Kconfig.sam_gmac: Fix queue definitions

Add missing queue entries for sam gmac. This update the queue selection
to proper handle all supported SAM SoC that uses GMAC driver.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: eth: eth_sam_gmac: Fix priority queues

The Atmel SAM SoC with ethernet port uses same GMAC driver. However,
there are differences between SoC GMAC implementation. Some SoCs have
priority queue and system can configure 0 up to 5, depending of SoC
version. This update current GMAC driver adding missing definitions.

Co-authored-by: Stephanos Ioannidis <root@stephanos.io>
Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1299,6 +1350,15 @@ static int priority2queue(enum net_priority priority)
+		0, 0, 0, 0, 1, 2, 3, 4
+#endif
+#if CONFIG_ETH_SAM_GMAC_QUEUES == 6
+		0, 0, 0, 1, 2, 3, 4, 5
```
I'm not complete sure about this, just add what seems to be a good start.

galak commented on drivers/ethernet/Kconfig.sam_gmac:
```diff
@@ -22,7 +22,12 @@ config ETH_SAM_GMAC_NAME
+		     !SOC_ATMEL_SAME70_REVB && \
+		     !SOC_ATMEL_SAMV71_REVB
+	range 1 6 if SOC_ATMEL_SAME70_REVB || \
+		     SOC_ATMEL_SAMV71_REVB
```
This seems like the kinda thing we should encode in the device tree.

stephanosio replied:
@galak Agreed.

@nandojve On second thought, since not initialising all priority queues create DMA problems and this is therefore not really "configurable", device tree would be a better place for this to go in.

nandojve replied:
> @nandojve On second thought, since not initialising all priority queues create DMA problems and this is therefore not really "configurable", device tree would be a better place for this to go in.

This configure the number of HW queues that app will use. The intention is save memory for small applications and it's a SW configuration.

The DMA problem was fixed by right definition on  **drivers/ethernet/eth_sam_gmac_priv.h** by @dgloeck. On top of it, I complemented to make sure devices with 1 queue will work too. Now, system initialize all queues always but properly select the active and inactive ones.

My intention was not create anything new, just fix! I'm open to do changes here if necessary.

nandojve replied:
> This seems like the kinda thing we should encode in the device tree.

I'm not complete aware about DT power on Zephyr but my first thought was: we need diferentiate variations A from B, we will duplicate all DT definitions to add this entry, that entry will not be used for any code generation and we need that information on Kconfig. The user will not be allowed to change it and this will be a long path just to drop 3 range definitions that already uses symbols from Kconfig.

This is a WIP, so if we need to change it, we will do. In this particular case, I'll try to make sure we are on the right path because things are a little _gray_ for me if we need use DT here.

What is the best in this case, keep it or move to DT?

stephanosio replied:
> The DMA problem was fixed by right definition on **drivers/ethernet/eth_sam_gmac_priv.h** by @dgloeck. On top of it, I complemented to make sure devices with 1 queue will work too. Now, system initialize all queues always but properly select the active and inactive ones.

In that case, this is indeed configurable and should stay as a Kconfig symbol.

Ideally, the device tree should specify the (maximum) number of queues supported in the hardware (i.e. `GMAC_QUEUE_NUM`), and the `ETH_SAM_GMAC_QUEUES` Kconfig symbol should specify the number of queues activated/used.

Noting that there is currently no GMAC definition in the dts for some reason, I suppose we could keep this Kconfig symbol with the emphasis that it refers to the "number of **active** hardware TX and RX queues," and add GMAC DT binding/instances later in a separate PR.

nandojve replied:
Sounds perfect! May I suggest create a enhanced. So, we can step back think the best way to do it. Maybe can it be done for 2.3? So, we fix now but get compromisse to solve for next release. You can assign it to me. We don't need to be in rush let's just focus on bug fix.

What do you think?

stephanosio replied:
@nandojve I have opened the issue #22997 for adding the GMAC device tree definition.

I agree that we should focus on the bugfix for now and try to get this PR in for 2.2.

stephanosio commented on drivers/ethernet/Kconfig.sam_gmac:
```diff
@@ -22,7 +22,12 @@ config ETH_SAM_GMAC_NAME
 config ETH_SAM_GMAC_QUEUES
 	int "Number of hardware TX and RX queues"
```
This Kconfig symbol specifies the "Number of **active** hardware TX and RX queues".

The maximum number of hardware TX and RX queues supported should be specified in the device tree.

stephanosio commented on drivers/ethernet/Kconfig.sam_gmac:
```diff
@@ -22,7 +22,12 @@ config ETH_SAM_GMAC_NAME
+		     !SOC_ATMEL_SAME70_REVB && \
+		     !SOC_ATMEL_SAMV71_REVB
+	range 1 6 if SOC_ATMEL_SAME70_REVB || \
+		     SOC_ATMEL_SAMV71_REVB
```
When the DT adds the value for the maximum number of hardware TX and RX queues supported in the future, that value should be used to specify the range.

stephanosio commented on drivers/ethernet/eth_sam_gmac_priv.h:
```diff
@@ -31,7 +31,17 @@
+#elif !defined(CONFIG_SOC_SERIES_SAM4E)
+#define GMAC_QUEUE_NUM                    3
+#else
+#define GMAC_QUEUE_NUM                    1
```
This value should eventually be derived from the device tree.

nandojve commented:
The work with queues is done! The remaining work related to DT will be addressed on #22997 and SAM4E fix and activation on #23031 just waiting this PR be merged.

stephanosio reviewed (approved):
Verified working on SAM E70 Rev. B.

mnkp commented on drivers/ethernet/eth_sam_gmac_priv.h:
```diff
@@ -31,7 +31,19 @@
+		 "GMAC_QUEUE_NUM doesn't match soc header");
+#endif
 /** Number of priority queues used */
 #define GMAC_PRIORITY_QUEUE_NO            (CONFIG_ETH_SAM_GMAC_QUEUES - 1)
```
We renamed `GMAC_QUEUE_NO` to `GMAC_QUEUE_NUM`. We probably should also rename `GMAC_PRIORITY_QUEUE_NO` to `GMAC_PRIORITY_QUEUE_NUM`.

stephanosio replied:
This naming scheme is quite confusing indeed.

`GMAC_QUEUE_NUM` and `GMAC_QUEUE_PRI_NUM` refer to "the number of (priority) queues supported in hardware", whereas `GMAC_PRIORITY_QUEUE_NO` refers to "the number of active/used (priority) queues."

I think we should just rename `GMAC_PRIORITY_QUEUE_NO` to something like `GMAC_ACTIVE_PRIORITY_QUEUE_NUM` and define `GMAC_ACTIVE_QUEUE_NUM` as `CONFIG_ETH_SAM_GMAC_QUEUES` and use that in place of it as well.

```
#define GMAC_QUEUE_NUM                    ...
#define GMAC_PRIORITY_QUEUE_NUM           (GMAC_QUEUE_NUM - 1)

#define GMAC_ACTIVE_QUEUE_NUM             CONFIG_ETH_SAM_GMAC_QUEUES 
#define GMAC_ACTIVE_PRIORITY_QUEUE_NUM    (GMAC_ACTIVE_QUEUE_NUM - 1)
```

nandojve replied:
Used @stephanosio suggestion.

mnkp commented on drivers/ethernet/eth_sam_gmac_priv.h:
```diff
@@ -142,11 +178,30 @@
+	GMAC_QUE_4,  /** Priority queue 4 */
+	GMAC_QUE_5,  /** Priority queue 5 */
+#endif
 };
```
This enum might have outlived its usefulness. The only place where it's used now is the initialization of `queue_list`, as in
```
	.que_idx = GMAC_QUE_3,
```
Here, an integer literal would be just as readable.

nandojve replied:
I prefer keep queue_idx for now but removed all preprocessor definitions. There are places that uses GMAC_QUE_1 and GMAC_QUE_2 and maybe we need review it. I took liberty to add some GMAC_QUE_x uses. Anyway, if will be need drop it I will not complain.

mnkp reviewed (changes requested):
LGTM, I do have however some comments related to style.

stephanosio commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -75,49 +75,243 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 /* RX descriptors list */
 static struct gmac_desc rx_desc_que0[MAIN_QUEUE_RX_DESC_COUNT]
 	__nocache __aligned(GMAC_DESC_ALIGNMENT);
+#if GMAC_QUEUE_NUM > 1
```
I have already done this in #23045, but we should handle all queue number variations here (i.e. `GMAC_QUEUE_NUM` = 3, 5, 6).

Although such a device does not exist at this time (as far as we know, that is), there could be one in the future and the current implementation will cause an insane behaviour in that case).

stephanosio replied:
One more thing: some declaration ifdefs use, "**total** queue count" and others use "**priority** queue count." I suppose this is because we only have "priority queue count" for the active.

Once we addressed the define naming issue (https://github.com/zephyrproject-rtos/zephyr/pull/22961/files#r383596823), we should use only one variant- preferably, the total queue count.

nandojve replied:
Add and "Co-authored-by: Stephanos Ioannidis <root@stephanos.io>"

stephanosio commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -75,49 +75,243 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#endif
+#if GMAC_QUEUE_NUM > 3
+static struct gmac_desc rx_desc_que3[PRIORITY_QUEUE3_RX_DESC_COUNT]
+	__nocache __aligned(GMAC_DESC_ALIGNMENT);
```
I believe that is being addressed in #23031 (in particular, https://github.com/zephyrproject-rtos/zephyr/pull/23031/commits/5e4074aae800e5f277d7a205382aef3d2e7fbecb).

Since there is currently no SAM 4E GMAC support (without #23031), this should not be an issue as far as this PR (bugfix) is concerned.

nandojve replied:
I decided to move any non priority queue related change to #23031

stephanosio commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -968,16 +1168,12 @@ static int nonpriority_queue_init(Gmac *gmac, struct gmac_queue *queue)
+		/* Attempt to use INCR4 AHB bursts (Default) */
+		GMAC_DCFGR_FBLDO_INCR4 |
+		/* DMA Queue Flags */
+		GMAC_DMA_QUEUE_FLAGS;
```
I suppose this is SAM 4E-related change; in that case, it should be part of #23031- not this bugfix PR.

nandojve replied:
It was dropped and will be added to #23031.

stephanosio commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -2050,6 +2201,85 @@ static struct eth_sam_dev_data eth0_data = {
+#endif
+#endif
+#if defined(CONFIG_SOC_ATMEL_SAME70_REVB) || \
+	defined(CONFIG_SOC_ATMEL_SAMV71_REVB)
```
Same here, we should handle all queue count variations. I have already done that in #23045.
https://github.com/zephyrproject-rtos/zephyr/pull/23045/commits/e5698aaf84051458945e2a680162a2fa1cda913b#diff-4b92059fe54cca181f5572c19a36712cR2169

Once we fix the define names, we should use `GMAC_ACTIVE_QUEUE_NUM` (=`CONFIG_ETH_SAM_GMAC_QUEUES`) here as well.

nandojve replied:
Add and "Co-authored-by: Stephanos Ioannidis root@stephanos.io"

nandojve commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -991,100 +1187,6 @@ static int nonpriority_queue_init(Gmac *gmac, struct gmac_queue *queue)
-		return priority_queue_init_as_idle(gmac, queue);
-	}
-}
-
```
The block was moved to allow drop many **defines** on code. As suggested, it was separated from others changes.

nandojve committed:
drivers: eth: eth_sam_gmac: Move queue init block

Move queue init block to avoid add many defines on code.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: eth: sam_gmac: Add priority queue checks

Improve priority queue conditional build. Now priority queue code is
enabled only if device have support to it. This enables GMAC driver
for devices with only one queue for RX/TX.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

mnkp commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -857,12 +1071,10 @@ static int gmac_init(Gmac *gmac, u32_t gmac_ncfgr_val)
 	(void)gmac->GMAC_ISR;
-	(void)gmac->GMAC_ISRPQ[GMAC_QUE_1 - 1];
-	(void)gmac->GMAC_ISRPQ[GMAC_QUE_2 - 1];
+	disable_all_queue_interrupt(gmac);
```
The name of the function should actually be `disable_all_priority_queue_interrupt` since 
```
	gmac->GMAC_IDR = UINT32_MAX;
	(void)gmac->GMAC_ISR;
```
is not included in `disable_all_queue_interrupt`.

nandojve replied:
Fixed on 821170d

mnkp commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -75,49 +75,269 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+{
+	u32_t idx;
+
+	for (idx = GMAC_QUE_0; idx < GMAC_PRIORITY_QUEUE_NUM; idx++) {
```
We shouldn't use `GMAC_QUE_0` as it is not a priority queue. Better to simply use `0`.

nandojve replied:
Good catch thank you! Fixed on 821170d!

mnkp reviewed (approved):
Sorry for the late review. I have two minor non-blocking comments. I won't be able to re-review the code in the next few days in case there are changes.

stephanosio commented:
@jhedberg Could we merge this for 2.2? Without this patch, the revision B SAM7 series devices (which all newly produced SAM7s are) cannot transmit Ethernet packets.

jhedberg commented:
> @jhedberg Could we merge this for 2.2? Without this patch, the revision B SAM7 series devices (which all newly produced SAM7s are) cannot transmit Ethernet packets.

Could you explain first what's going on with the linked issues. Both of them are closed, so exactly what GitHub issue is this fixing?

stephanosio commented:
> > @jhedberg Could we merge this for 2.2? Without this patch, the revision B SAM7 series devices (which all newly produced SAM7s are) cannot transmit Ethernet packets.
> 
> Could you explain first what's going on with the linked issues. Both of them are closed, so exactly what GitHub issue is this fixing?

@jhedberg Those issues should not have been closed, as the problem described there was not fixed (though both issues were more of a "question" that a proper "bug report").

This PR fixes the problem reported in #22890 and #23035.
